src/polyswarm_api/endpoint.py

```diff
@@ -0,0 +1,254 @@
+        req = {
+            'method': 'GET',
+            'url': '{}/historical/results'.format(self.hunt_base),
+            'params': {
+                'with_bounty_results': utils.bool_to_int[with_bounty_results],
+                'with_metadata': utils.bool_to_int[with_metadata],
+                'limit': limit,
+                'offset': offset,
+            },
+        }
+
+        if id:
+            req['params']['id'] = id
+
+        return req
+
+    def historical_delete(self, hunt_id):
+        return {
+            'method': 'DELETE',
+            'url': '{}/historical'.format(self.hunt_base),
+            'json': {'hunt_id': hunt_id}
+        }
+
+    def live_delete(self, hunt_id):
+        return {
+            'method': 'DELETE',
+            'url': '{}/live'.format(self.hunt_base),
+            'json': {'hunt_id': hunt_id}
+        }
+
+    def historical_list(self):
+        return {
+            'method': 'GET',
+            'url': '{}/historical'.format(self.hunt_base),
+            'params': {'all': 'true'},
+        }
+
+    def live_list(self):
+        return {
+            'method': 'GET',
+            'url': '{}/live'.format(self.hunt_base),
+            'params': {'all': 'true'},
+        }
+
+
+class PolyswarmRequestExecutor(object):
+    """ This class accepts requests from a PolyswarmRequestGenerator and executes it """
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES, request_cls=None):
+        self.session = request_cls(key, retries)
+        self.unauth_session = request_cls(key=None, retries=retries)
+        self.timeout = timeout
+
+    def execute(self, request):
+        if 'timeout' not in request:
+            request['timeout'] = self.timeout
+        return self.session.request(**request)
+
+    def unauth_execute(self, request):
+        return self.unauth_session.request(**request)
+
+    def _get_engine_names(self, request):
+        return self.unauth_session.request(**request)
+
+    def __getattr__(self, name):
+        if name in self.__dict__:
+            return self.__dict__[name]
+        return self.execute
+
+    def _download_to_fh(self, req, fh):
+        raise NotImplementedError
+
+    def download(self, request):
+        request, fh = request
+        if isinstance(fh, str):
+            fh = open(fh, 'wb')
+
+        req = self.execute(request)
+        return self._download_to_fh(req, fh)
+
+    def download_archive(self, request):
+        request, fh = request
+        if isinstance(fh, str):
+            fh = open(fh, 'wb')
+
+        req = self.unauth_execute(request)
+        return self._download_to_fh(req, fh)
+
+
+class PolyswarmFuturesExecutor(PolyswarmRequestExecutor):
+    def __init__(self, key, timeout=const.DEFAULT_HTTP_TIMEOUT, retries=const.DEFAULT_RETRIES):
+        super(PolyswarmFuturesExecutor, self).__init__(key, timeout, retries, PolyswarmHTTPFutures)
```

Comment:
Once we get instances, these custom `__init__()` become obsolete and don't need to exist, we only care that the `HTTP` class implement the correct methods. 

P.S.: I kind of ended up making the comments out of order, read the big one down below and come back for this one. Sorry for that.